# Introduction

EPIHAPMPI is a parallel, distributed memory program designed for bi-allelic diploid species and is based on the quantitative genetics model [@Da2022;@Wang2014]. This model incorporates multifactorial effects (SNP, haplotype, and epistasis) to investigate the contributions of global low-order and local high-order epistasis effects to the phenotypic variance and the accuracy of genomic prediction of quantitative traits.
Users can select up to third-order epistatic effects, additive and dominance effects, and haplotype effects to analyze the undrlying genetic mechanisms. Analogous to EPIHAP (shared memory), this program is intended for large datasets that can not be computed on a regular computer or a single node; it is indeded to run on clusters and requires MPI.
This analysis provides genomic best linear unbiased prediction (GBLUP) with associated reliability for individuals with and without phenotypic observations, including a computationally efficient method for large validation populations, and genomic restricted maximum estimation (GREML) of the variance and associated heritability using a combination of EM-REML and AI-REML iterative algorithms.

## Single SNP effects
EPIHAPMPI can construct genetic relationship matrices (GRMs) for additive (A), dominance (D) effects [@Wang2014].

## Epistasis 

EPIHAPMPI can construct genetic relationship matrices (GRMs) for  additive × additive (AA), additive × dominance (AD), dominance × dominance (DD), additive × additive × additive (AAA), additive × additive × dominance (AAD), additive × dominance × dominance (ADD), dominance × dominance × dominance (DDD) effects using the Approximate Genomic Epistasis Relationship Matrices (AGERM) method (CITE: Henderson, 1985; Jiang and Reif, 2020)[@Da2022].

## Haplotype

User defined haplotype blocks can incorporate structural and function information for genomic prediction and estimation, but this method requires large amounts of data preparation due to large number of candidate models. EPIHAPMPI, and the accompanying pipeline programs provide capability to analyze many large-scale haplotype models and automate the process.
EPIHAPMPI implements a multi-allelic haplotype mixed model that treats each haplotype block as a ‘locus’ and each haplotype within the haplotype block as an ‘allele’ [@Da2022], based on the GVCHAP pipeline [@gvchap2020] and uses the GREML_CE method described in [@Wang2014].

## Combination
EPIHAPMPI can combine effects of single SNP, epistasis, and haplotype models to estimate variance components and the associated heritability.
