# Running EPIHAPMPI

## Creating GRMs

There are 3 orders of Genomic Relationship Matrices that can be made using this software:

1. Single order, conventional GRM
- SNP Additive
- SNP Dominance
- Haplotype Additive

2. Second order SNP epistasis GRM
- (AxA) Additive by Additive
- (AxD) Additive by Dominance
- (DxD) Dominance by Dominance

3. Third order SNP epistasis GRM
- (AxAxA) Additive by Additive by Additive
- (AxAxD) Additive by Additive by Dominance
- (AxDxD) Additive by Dominance by Dominance
- (DxDxD) Dominance by Dominance by Dominance

## Running full models

Pick the effects you want to include in your model.

For additive and dominance snp effects only:
```bash
# replace 4 with the number of MPI tasks you have available

mpirun -np 4 ./bin/grm.snp -o mygrms chr1.dat chr2.dat chr3.dat -a -d

# or if all chromosomes in one file:
mpirun -np 4 ./bin/grm.snp -o mygrms all.dat -a -d

# Output:
#  mygrms.g.A
#  mygrms.g.D
```

For all effects up to second order excluding dominance:
```bash
# replace 4 with the number of MPI tasks you have available

mpirun -np 4 ./bin/grm.snp -o mygrms chr1.dat chr2.dat chr3.dat --a --aa
# or if all chromosomes in one file:
mpirun -np 4 ./bin/grm.snp -o mygrms all.dat --a --aa

# Output:
#  mygrms.g.A
#  mygrms.g.AA
```

Or for all effects up to second order:
```bash
# replace 4 with the number of MPI tasks you have available

mpirun -np 4 ./bin/grm.snp -o out/mygrms chr1.dat chr2.dat chr3.dat
# or if all chromosomes in one file:
mpirun -np 4 ./bin/grm.snp -o out/mygrms all.dat

# Output:
#  out/mygrms.g.A
#  out/mygrms.g.D
#  out/mygrms.g.AA
#  out/mygrms.g.AD
#  out/mygrms.g.DD
```

For haplotype effects:
```bash
# replace 4 with the number of MPI tasks you have available

mpirun -np 4 ./bin/grm.hap chr1 chr2 chr3 -o mygrms
# or if all chromosomes in one file:
mpirun -np 4 ./bin/grm.hap -o out/mygrms all.hap

# Output:
#  mygrms.g.AH
```

For all effects up to third order:
```bash
# uses already created first and second order GRMs under prefix "out/mygrms"
./bin/grm.snp.3order --threads 8 --buffer 128M out/mygrms

# Output:
#  out/mygrms.g.AAA
#  out/mygrms.g.AAD
#  out/mygrms.g.ADD
#  out/mygrms.g.DDD
```

For just AAA (AxAxA) effects:
```bash
# uses already created first and second order GRMs under prefix "out/mygrms"
# to create additive by additive by additive GRM
./bin/grm.snp.3order --threads 16 --aaa out/mygrms
# or all possible third order GRMs
./bin/grm.snp.3order --threads 16 out/mygrms

# Output:
#  out/mygrms.g.AAA
```

Then run the perdiction model using the effects you want to include by
specifying the variance componentes. Residual variance component is always included
and must be specified:
```bash
# run the prediction using GRMs
# for snp additive, snp additive by addive, and haplotype additive
# in the reml model
# replace 4 with the number of MPI tasks you have available
mpirun -np 4 ./bin/reml \
    --input out/mygrms \
    --pheno 3,pheotype.txt \
    --va 2.435 \
    --vaa 3.12 \
    --vah 1.349E-01 \
    --out a.d.ah.model
```
